src/System.Memory/src/System/SpanHelpers.BinarySearch.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    internal static partial class SpanHelpers
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static int BinarySearch<T, TComparable>(
+            this ReadOnlySpan<T> span, TComparable comparable)
+            where TComparable : IComparable<T>
+        {
+            if (comparable == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.comparable);
+            // TODO: Make `ref readonly`/`in` when Unsafe.Add(ReadOnly) supports it
+            return BinarySearch(ref span.DangerousGetPinnableReference(), span.Length, comparable);
+        }
+
+        // TODO: Make s `ref readonly`/`in` when Unsafe.Add(ReadOnly) supports it
+        internal static int BinarySearch<T, TComparable>(
+            ref T s, int length, TComparable comparable) 
+            where TComparable : IComparable<T>
+        {
+            // Array.BinarySearch implementation:
+            // https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Array.cs#L802
+            int lo = 0;
+            int hi = length - 1;
+            while (lo <= hi)
+            {
+                // lo or hi will never be negative inside the loop
```

Comment:
If span.Length == 0, hi will be negative, no? Does this require a comment update?